### PR TITLE
Masek-Paterson edit distance via Four Russians method

### PR DESCRIPTION
**Pull request by kasiabulat**
Implementation and tests for the algorithm for computing edit distances and longest common subsequence using the method of "Four Russians". Based on the paper ["Masek, Paterson - Faster Algorithm Computing String Edit Distances"](https://www.sciencedirect.com/science/article/pii/0022000080900021#:~:text=A%20faster%20algorithm%20computing%20string%20edit%20distances%E2%98%86&text=The%20edit%20distance%20between%20two,one%20string%20into%20the%20other.&text=These%20conditions%20are%20necessary%20for%20the%20algorithm%20to%20achieve%20the%20time%20bound.).

### CODE REVIEW

File: approximate_string_matching/four_russians_helpers.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+    sub_length_results = [[] for _ in range(m + 1)]
+    sub_length_results[0] = []
+    sub_length_results[1] = A
+
+    for size in range(2, m + 1):
+      sub_length_results[size] = []
+      for left_substring in sub_length_results[1]:
+        for right_substring in sub_length_results[size - 1]:
+          sub_length_results[size].append(left_substring + right_substring)
```
Comment (krzysztof-turowski) [inline approximate_string_matching/four_russians_helpers.py:57]: More Pythonic would be something like
sub_length_results[size] = [left_substring + right_substring for left_substring in sub_length_results[1] for right_substring in sub_length_results[size - 1]]

File: text/four-russians.tex [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+\noindent Aby tego dokonać, musimy być w stanie wyenumerować wszystkie możliwe macierze rozmiaru $m$. Zakładamy, że alfabet jest skończony, więc słów długości $m$ jest skończona ilość. Jednak generowanie wszystkich możliwych wektorów inicjalnych może zająć zbyt dużo czasu - wartości komórek mają tendencję do wzrostu wraz z rozszerzaniem się macierzy głównej. Kiedy jednak ograniczymy funkcję kosztu do przypisywania operacjom edycyjnym wyłącznie wielokrotności jakiejś liczby rzeczywistej, okaże się, że mamy tylko skończoną ilość różnic pomiędzy sąsiednimi komórkami w macierzy edycyjnej.\\\\
```
Comment (krzysztof-turowski) [inline text/four-russians.tex:46]: wyenumerować -> wygenerować
słów długości $m$ jest skończona ilość -> jest $|\A|^m$
skończoną ilość różnic -> skończoną liczbę różnic ("ilość" is in principle for uncountable stuff)

File: approximate_string_matching/four_russians_helpers.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+      lcs_part, i, j = self.restore_lcs_part(
+          C, D,
+          [0] + P[I][J - 1],
+          [0] + Q[I - 1][J],
+          m, i, j)
+
+      if i == 0:
+        I = I - 1
+        i = m
```
Comment (krzysztof-turowski) [inline approximate_string_matching/four_russians_helpers.py:231]: You may write it in one line: I, i = I - 1, m. Similarly below.

File: approximate_string_matching/four_russians_helpers.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+  @staticmethod
+  def get_kth_substring(k, m, text_1):
+    return text_1[((k - 1) * m + 1):(k * m + 1)]
+
+  def restore_lcs(self, text_1, text_2, P, Q, m):
+    lcs = ""
+
+    # indices on the matrix of submatrices
+    I = int((len(text_1) - 1) / m)
+    J = int((len(text_2) - 1) / m)
```
Comment (krzysztof-turowski) [inline approximate_string_matching/four_russians_helpers.py:213]: You may write it in one line as a multiple assignment. And if you want integer division, that's x // y in Python.

File: approximate_string_matching/four_russians_helpers.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    S = tuple(S)
+
+    if C not in storage:
+      storage[C] = {}
+    if D not in storage[C]:
+      storage[C][D] = {}
+    if R not in storage[C][D]:
+      storage[C][D][R] = {}
+
+    storage[C][D][R][S] = (R_new, S_new)
```
Comment (krzysztof-turowski) [inline approximate_string_matching/four_russians_helpers.py:74]: Well, it's a matter of preference, but I'd either use defaultdict not to have to initialize missing entries or even go for storage[(C, D, R, S)], so with the tuple as a key.

File: approximate_string_matching/four_russians_helpers.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        )
+
+  def algorithm_y(self, m, A, step_size_bound):
+    """ Preprocesses data by creating helper submatrices """
+
+    if len(A) == 0:
+      return []
+
+    A = ['#'] + A
+    strings = self.get_all_strings(m, A, '#')
```
Comment (krzysztof-turowski) [inline approximate_string_matching/four_russians_helpers.py:127]: I'm wondering whether all this handling of '#' may be discarded as it adds somewhat complication to your code. Just in the main function we may use text1[1:] and text2[1:] with helper containing no '#' symbol.

File: approximate_string_matching/four_russians_helpers.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+            self.delete_cost_function(C[i]) + M[i - 1][j],
+            self.insert_cost_function(D[j]) + M[i][j - 1]
+        )
+    return M
+
+  def restore_lcs_part(self, C, D, R, S, m, i1, j1):
+    M = self.restore_matrix(C, D, R, S, m)
+
+    i = i1
+    j = j1
```
Comment (krzysztof-turowski) [inline approximate_string_matching/four_russians_helpers.py:188]: A natural place for a multiple assignment. Also it may be applied in many other places.

**Comment by kasiabulat** [comment]
Thank you for the feedback and appreciation! Implementing the algorithm was indeed a bit annoying in some places. I have a quite busy time now, so, unfortunately, I won't be able to add fixes.

File: approximate_string_matching/four_russians_helpers.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
+import math
+
+lcs_delete_cost_function = lambda a: 1
+lcs_insert_cost_function = lambda b: 1
+lcs_substitute_cost_function = lambda a, b: 0 if (a == b) else 2
```
[the remark this answers is not in the document]
Comment (kasiabulat) [reply approximate_string_matching/four_russians_helpers.py:5]: I actually didn't realize that there was a ScoreMatrix tuple and I can use it here...

**Comment by krzysztof-turowski** [comment]
> Thank you for the feedback and appreciation! Implementing the algorithm was indeed a bit annoying in some places. I have a quite busy time now, so, unfortunately, I won't be able to add fixes.

Thank you for letting me know, I made small style changes in places and I'll leave the major changes until I'll be merging all your contributions.

**Pull request merged**